### ISSUE TITLE
`SoftBody3D` crashes when colliding with `HeightMapShape3D`

### ISSUE THREAD

**Issue by jordanmichaelrushing** [godot-jolt/godot-jolt#951]
Godot: 4.3
Jolt: 0.13.0
OS: Win 11
Renderer: Forward+
Vulkan: 1.3.277
GPU: AMD Radeon RX 7900 GRE

Describing Issue: After going through 2d menus to start the game, the 3d scene attempts to load but then Windows brings up the "Godot Engine has stopped working" popup.

The logs aren't showing any problem when the issue occurs, so I don't really have anything to show (if someone knows how to dump a log that isn't normally shown, that might be helpful). The only thing I have noticed is when the project first loads I receive an error:

```
ERROR: File 'C:\Users\Jordan\Desktop\Element Racers - Veggie Vendetta\addons\godot-jolt\windows/godot-jolt_windows-x64_editor.pdb' does not exist.
```

Do I need to delete the .godot folder since I'm upgrading? It's also worth calling out that the reason I'm posting this here instead of on Godot's GH is because when I go with the Engine Default, the 3d game loads fine. This also worked great prior to this recent upgrade, so something seems to have broke, and I'm not getting any logs in engine or in the console. Thanks for any help anyone can give!

**Comment by jordanmichaelrushing** [comment]
Update: I think it may have to do with a soft body I added. I removed the soft body plane from the main 3d scene and now it loads. I'll try and give it a delayed load to see if it borks the  engine again, but this time with an issue printed in the logs

**Comment by jordanmichaelrushing** [comment]
Confirmed, it's the softbody. I went with an out of the box softbody3d, with several different shapes and it crashes every time. Is there something I need to do to get softbody3d working in this latest version?

**Comment by mihe** [comment]
That's not good. You shouldn't have to do anything special to make it work, no.

Would you be able to share a repro project of some kind, so that I can debug it on my end, @jordanmichaelrushing?

If that isn't an option, please download [`godot-jolt_v0.13.0-stable_symbols.zip`](https://github.com/godot-jolt/godot-jolt/releases/download/v0.13.0-stable/godot-jolt_v0.13.0-stable_symbols.zip) (found [here](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.13.0-stable)) and extract them to your project, so that `addons/godot-jolt/windows/godot-jolt_windows-x64.pdb` sits next to `addons/godot-jolt/windows/godot-jolt_windows-x64.dll`. Then you need to run the `Godot_v4.3-stable_win64_console.exe` executable that comes with the official Godot release, which will open a console window along with the editor when you start it. With this setup you should get a crash backtrace logged in that command-line window when something crashes, which you can then paste here.

**Comment by jordanmichaelrushing** [comment]
Sure, give me like 5 hours and I can tackle this! Thanks @mihe !

**Comment by mihe** [comment]
Just a friendly ping here, @jordanmichaelrushing, in case you have time to do a quick test with the debug symbols I linked above.

**Comment by mihe** [comment]
I obviously can't know if #958 actually fixes this, but I did find one potential crash bug (although the use-case was unlikely) and revamped some problematic error-handling in the process. I'm hoping that may have fixed your issue too.

If you do get the chance to try out the debug symbols above, please share the backtrace, and we can confirm whether it did end up being fixed by #958.

**Comment by jordanmichaelrushing** [comment]
Hey so sorry, I got really sick. Coming back to this tonight though and will let you know

**Comment by jordanmichaelrushing** [comment]
@mihe Can I give you temp access to the game repo? I added everything and still just getting a crash without any information added in the logs

**Comment by mihe** [comment]
Sure, that's no problem, so long as there's a solid repro.

**Comment by jordanmichaelrushing** [comment]
@mihe added you to the org. Please let me know if you have any questions. Happy to help out as much as I can here

**Comment by mihe** [comment]
I seem to be able to reproduce the crash.

It's crashing in one of Jolt's worker threads, as it's doing the collision detection between the `SoftBody3D` and one of the `HeightMapShape3D` created by Terrain3D.

I get the same crash if I extract a Jolt snapshot of the scene and run it in Jolt's own `Samples` application, and I'm not seeing any sort of Jolt asserts being triggered during the setup of the shapes, so I suspect this might be a bug in Jolt.

@jrouwe Before I ask the author here for permission to share the snapshot, do these callstacks give you enough information to see what the cause of the crash might be perhaps? The checked out commit is jrouwe/JoltPhysics@9320e5d6c9025f21fe5367ec15303947084a5953.

It's jumping between [this assert](https://github.com/jrouwe/JoltPhysics/blob/9320e5d6c9025f21fe5367ec15303947084a5953/Jolt/Physics/Collision/Shape/HeightFieldShape.cpp#L768), with this callstack, where `GetNumBlocks()` is 513 due to `mSampleCount` being 1026 and `mBlockSize` being 2, and `inBlockX` is 513 and `inBlockY` is 511.

```
JPH::HeightFieldShape::GetBlockOffsetAndScale(unsigned int inBlockX, unsigned int inBlockY, unsigned int inRangeBlockOffset, unsigned int inRangeBlockStride, float & outBlockOffset, float & outBlockScale) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:768)
JPH::HeightFieldShape::DecodingContext::WalkHeightField<`JPH::HeightFieldShape::CollideSoftBodyVertices'::`2'::Visitor>(JPH::HeightFieldShape::CollideSoftBodyVertices::__l2::Visitor & ioVisitor) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:1777)
JPH::HeightFieldShape::WalkHeightField<`JPH::HeightFieldShape::CollideSoftBodyVertices'::`2'::Visitor>(JPH::HeightFieldShape::CollideSoftBodyVertices::__l2::Visitor & ioVisitor) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:2057)
JPH::HeightFieldShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:2242)
JPH::ScaledShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\ScaledShape.cpp:144)
JPH::CompoundShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\CompoundShape.cpp:270)
JPH::SoftBodyMotionProperties::DetermineCollisionPlanes(const JPH::SoftBodyUpdateContext & inContext, unsigned int inVertexStart, unsigned int inNumVertices) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:201)
JPH::SoftBodyMotionProperties::ParallelDetermineCollisionPlanes(JPH::SoftBodyUpdateContext & ioContext) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:793)
JPH::SoftBodyMotionProperties::ParallelUpdate(JPH::SoftBodyUpdateContext & ioContext, const JPH::PhysicsSettings & inPhysicsSettings) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:896)
JPH::PhysicsSystem::JobSoftBodySimulate(JPH::PhysicsUpdateContext * ioContext, unsigned int inThreadIndex) (Jolt\Physics\PhysicsSystem.cpp:2561)
JPH::PhysicsSystem::JobSoftBodyPrepare::__l11::<lambda>() (Jolt\Physics\PhysicsSystem.cpp:2496)
[...]
JPH::JobSystem::Job::Execute() (Jolt\Core\JobSystem.h:240)
JPH::JobSystemThreadPool::ThreadMain(int inThreadIndex) (Jolt\Core\JobSystemThreadPool.cpp:348)
```

... and [this assert](https://github.com/jrouwe/JoltPhysics/blob/9320e5d6c9025f21fe5367ec15303947084a5953/Jolt/Physics/Collision/Shape/HeightFieldShape.cpp#L785), with this callstack, where again `mSampleCount` is 1026, `inX` is 1026 and `inY` is 1022:

```
JPH::HeightFieldShape::GetHeightSample(unsigned int inX, unsigned int inY) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:785)
JPH::HeightFieldShape::GetPosition(unsigned int inX, unsigned int inY, float inBlockOffset, float inBlockScale, bool & outNoCollision) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:803)
JPH::HeightFieldShape::DecodingContext::WalkHeightField<`JPH::HeightFieldShape::CollideSoftBodyVertices'::`2'::Visitor>(JPH::HeightFieldShape::CollideSoftBodyVertices::__l2::Visitor & ioVisitor) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:1818)
JPH::HeightFieldShape::WalkHeightField<`JPH::HeightFieldShape::CollideSoftBodyVertices'::`2'::Visitor>(JPH::HeightFieldShape::CollideSoftBodyVertices::__l2::Visitor & ioVisitor) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:2057)
JPH::HeightFieldShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:2242)
JPH::ScaledShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\ScaledShape.cpp:144)
JPH::CompoundShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\CompoundShape.cpp:270)
JPH::SoftBodyMotionProperties::DetermineCollisionPlanes(const JPH::SoftBodyUpdateContext & inContext, unsigned int inVertexStart, unsigned int inNumVertices) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:201)
JPH::SoftBodyMotionProperties::ParallelDetermineCollisionPlanes(JPH::SoftBodyUpdateContext & ioContext) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:793)
JPH::SoftBodyMotionProperties::ParallelUpdate(JPH::SoftBodyUpdateContext & ioContext, const JPH::PhysicsSettings & inPhysicsSettings) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:896)
JPH::PhysicsSystem::JobSoftBodySimulate(JPH::PhysicsUpdateContext * ioContext, unsigned int inThreadIndex) (Jolt\Physics\PhysicsSystem.cpp:2561)
JPH::PhysicsSystem::JobSoftBodyPrepare::__l11::<lambda>() (Jolt\Physics\PhysicsSystem.cpp:2496)
[...]
JPH::JobSystem::Job::Execute() (Jolt\Core\JobSystem.h:240)
JPH::JobSystemThreadPool::ThreadMain(int inThreadIndex) (Jolt\Core\JobSystemThreadPool.cpp:348)](https://github.com/jrouwe/JoltPhysics/blob/9320e5d6c9025f21fe5367ec15303947084a5953/Jolt/Physics/Collision/Shape/HeightFieldShape.cpp#L785)
```

... and if I just let it run it seems to crash [here](https://github.com/jrouwe/JoltPhysics/blob/9320e5d6c9025f21fe5367ec15303947084a5953/Jolt/Physics/Collision/Shape/HeightFieldShape.cpp#L796), with this callstack, due to `height_samples` pointing to invalid memory:

```
JPH::HeightFieldShape::GetHeightSample(unsigned int inX, unsigned int inY) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:796)
JPH::HeightFieldShape::GetPosition(unsigned int inX, unsigned int inY, float inBlockOffset, float inBlockScale, bool & outNoCollision) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:803)
JPH::HeightFieldShape::DecodingContext::WalkHeightField<`JPH::HeightFieldShape::CollideSoftBodyVertices'::`2'::Visitor>(JPH::HeightFieldShape::CollideSoftBodyVertices::__l2::Visitor & ioVisitor) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:1782)
JPH::HeightFieldShape::WalkHeightField<`JPH::HeightFieldShape::CollideSoftBodyVertices'::`2'::Visitor>(JPH::HeightFieldShape::CollideSoftBodyVertices::__l2::Visitor & ioVisitor) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:2057)
JPH::HeightFieldShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\HeightFieldShape.cpp:2242)
JPH::ScaledShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\ScaledShape.cpp:144)
JPH::CompoundShape::CollideSoftBodyVertices(const JPH::Mat44 & inCenterOfMassTransform, const JPH::Vec3 inScale, JPH::SoftBodyVertex * ioVertices, unsigned int inNumVertices, float inDeltaTime, const JPH::Vec3 inDisplacementDueToGravity, int inCollidingShapeIndex) (Jolt\Physics\Collision\Shape\CompoundShape.cpp:270)
JPH::SoftBodyMotionProperties::DetermineCollisionPlanes(const JPH::SoftBodyUpdateContext & inContext, unsigned int inVertexStart, unsigned int inNumVertices) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:201)
JPH::SoftBodyMotionProperties::ParallelDetermineCollisionPlanes(JPH::SoftBodyUpdateContext & ioContext) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:793)
JPH::SoftBodyMotionProperties::ParallelUpdate(JPH::SoftBodyUpdateContext & ioContext, const JPH::PhysicsSettings & inPhysicsSettings) (Jolt\Physics\SoftBody\SoftBodyMotionProperties.cpp:896)
JPH::PhysicsSystem::JobSoftBodySimulate(JPH::PhysicsUpdateContext * ioContext, unsigned int inThreadIndex) (Jolt\Physics\PhysicsSystem.cpp:2561)
JPH::PhysicsSystem::JobSoftBodyPrepare::__l11::<lambda>() (Jolt\Physics\PhysicsSystem.cpp:2496)
[...]
JPH::JobSystem::Job::Execute() (Jolt\Core\JobSystem.h:240)
JPH::JobSystemWithBarrier::BarrierImpl::Wait() (Jolt\Core\JobSystemWithBarrier.cpp:141)
JPH::JobSystemWithBarrier::WaitForJobs(JPH::JobSystem::Barrier * inBarrier) (Jolt\Core\JobSystemWithBarrier.cpp:225)
JPH::PhysicsSystem::Update(float inDeltaTime, int inCollisionSteps, JPH::TempAllocator * inTempAllocator, JPH::JobSystem * inJobSystem) (Jolt\Physics\PhysicsSystem.cpp:572)
SamplesApp::StepPhysics(JPH::JobSystem * inJobSystem) (Samples\SamplesApp.cpp:2435)
```

**Comment by jrouwe** [comment]
I can reproduce the asserts, I think it is due to mSampleCount / mBlockSize not being a power of 2. I'll see if I can create a fix.

It doesn't crash for me though, but I can see it reads very much out of bounds, so perhaps I'm just lucky that that memory is mapped.

**Comment by jrouwe** [comment]
Ok, I think it is fixed. @mihe can you integrate the latest version of Jolt?

**Comment by mihe** [comment]
I just tested both the snapshot and the repro project against latest Jolt and the crash does indeed seem to be resolved.

I'll update Jolt right away, thank you!

**Comment by mihe** [comment]
@jordanmichaelrushing I kicked myself from the GitHub organization you invited me to, and deleted the local copy of the project as well.

I'll try to cut a new release some time this weekend, if I can find the time. Until then you should be able to [build from source](https://github.com/godot-jolt/godot-jolt/blob/master/docs/building.md) yourself if you really need this fix.

**Comment by jordanmichaelrushing** [comment]
Hero. Thanks so much @mihe

**Comment by mihe** [comment]
You really should be thanking @jrouwe in this case. :)

**Comment by jordanmichaelrushing** [comment]
@mihe finally getting back from a burnout. I looked at the releases, looks like it's not been bumped, so I should build from source, ya?

**Comment by mihe** [comment]
@jordanmichaelrushing Yes. I had hoped to put out another release by now, but it'll be a while longer unfortunately.

**Comment by mihe** [comment]
Apologies for the long wait, but the fix for this can now be found in the newly released [0.14.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.14.0-stable) release.

Note that the release on Godot Asset Library is still pending approval, so you might need to download it from GitHub for the time being.
